cmd/storage-rest-client.go

```diff
@@ -0,0 +1,342 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"bytes"
+	"crypto/tls"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/url"
+	"path"
+	"strconv"
+
+	"encoding/gob"
+	"encoding/hex"
+
+	"github.com/minio/minio/cmd/logger"
+	"github.com/minio/minio/cmd/rest"
+	xnet "github.com/minio/minio/pkg/net"
+)
+
+func isNetworkDisconnectError(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	if uerr, isURLError := err.(*url.Error); isURLError {
+		if uerr.Timeout() {
+			return true
+		}
+
+		err = uerr.Err
+	}
+
+	_, isNetOpError := err.(*net.OpError)
+	return isNetOpError
+}
+
+// Converts rpc.ServerError to underlying error. This function is
+// written so that the storageAPI errors are consistent across network
+// disks as well.
+func toStorageErr(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	if isNetworkDisconnectError(err) {
+		return errDiskNotFound
+	}
+
+	switch err.Error() {
+	case io.EOF.Error():
+		return io.EOF
+	case io.ErrUnexpectedEOF.Error():
+		return io.ErrUnexpectedEOF
+	case errUnexpected.Error():
+		return errUnexpected
+	case errDiskFull.Error():
+		return errDiskFull
+	case errVolumeNotFound.Error():
+		return errVolumeNotFound
+	case errVolumeExists.Error():
+		return errVolumeExists
+	case errFileNotFound.Error():
+		return errFileNotFound
+	case errFileNameTooLong.Error():
+		return errFileNameTooLong
+	case errFileAccessDenied.Error():
+		return errFileAccessDenied
+	case errIsNotRegular.Error():
+		return errIsNotRegular
+	case errVolumeNotEmpty.Error():
+		return errVolumeNotEmpty
+	case errVolumeAccessDenied.Error():
+		return errVolumeAccessDenied
+	case errCorruptedFormat.Error():
+		return errCorruptedFormat
+	case errUnformattedDisk.Error():
+		return errUnformattedDisk
+	case errInvalidAccessKeyID.Error():
+		return errInvalidAccessKeyID
+	case errAuthentication.Error():
+		return errAuthentication
+	case errRPCAPIVersionUnsupported.Error():
+		return errRPCAPIVersionUnsupported
+	case errServerTimeMismatch.Error():
+		return errServerTimeMismatch
+	}
+	return err
+}
+
+// Abstracts a remote disk.
+type storageRESTClient struct {
+	endpoint   Endpoint
+	restClient *rest.Client
+	connected  bool
+	lastError  error
+}
+
+// Wrapper to restClient.Call to handle network errors, in case of network error the connection is makred disconnected
+// permanently. The only way to restore the storage connection is at the xl-sets layer by xlsets.monitorAndConnectEndpoints()
+// after verifying format.json
+func (client *storageRESTClient) call(method string, values url.Values, body io.Reader) (respBody io.ReadCloser, err error) {
+	if !client.connected {
+		return nil, errDiskNotFound
+	}
+	respBody, err = client.restClient.Call(method, values, body)
+	if err == nil {
+		return respBody, nil
+	}
+	if isNetworkDisconnectError(err) {
+		client.connected = false
+		client.lastError = err
```

Comment:
```
	respBody, err = client.restClient.Call(method, values, body)
        client.lastError =err
```

This will show the result of the last RPC call (probably we should pick a better name than lastError here)

Reply:
I see that LastError() is not useful, because a `disk` is never re-used after the first error. I will remove it in a different PR.

Reply:
> I see that LastError() is not useful, because a `disk` is never re-used after the first error. I will remove it in a different PR.

There is other errors that we need to detect other than network disconnect here, like authentication failed, certificate authority unknown, etc..

Reply:
got it